### PR TITLE
[WIP] investigating failing annotation test

### PR DESCRIPTION
**Pull request by chrishavlin**
Looking into #4785 

I wasn't able to reproduce the failure locally, so i'm adding an image test in this PR to be able to check out the image artifact to see if that helps...

### CODE REVIEW

**Comment by chrishavlin** [comment]
rebased off #4784

**Comment by chrishavlin** [comment]
the artifact image produced by CI matches what I get locally, so no insight there... latest push adjusts the existing test_annotation unit test to also check the image size before the annotation call.

**Comment by chrishavlin** [comment]
latest push disables the text annotation in the unit test (to isolate the text addition from the colorbar addition) and makes the text in the image comparison test more obvious.

**Comment by chrishavlin** [comment]
well i've got to leave this for a while. no real progress other than ruling out the text overlay. maybe something odd is going on when the colorbar axis is added?

**Comment by chrishavlin** [comment]
latest push adjusts the test workflow to only run the failing test (obviously don't merge this...)

**Comment by chrishavlin** [comment]
summary so far: running just the failed test in CI or the `yt/visualization` subset of tests passes in 3.12

**Comment by chrishavlin** [comment]
latest push strips down the test action to just run 3.12 on ubuntu (just trying to speed up the process of isolating the issue and lessen the load on github actions while I'm iterating here)

**Comment by chrishavlin** [comment]
oh interesting. now that i'm properly catching the failure, `test_annotation` does in fact pass when re-run after failing in the first call (with `pytest --lf`).

**Comment by chrishavlin** [comment]
latest push removes some of the slower to install deps (ones that require wheels to build for 3.12), this will slightly change the test suite, but probably will still fail... after which I'm going to start a bit more trial and error with different test subsets (unless anyone else has a better idea of how to isolate the issue! )

**Comment by chrishavlin** [comment]
oh hey, should try this pytest plugin... https://github.com/asottile/detect-test-pollution I'll keep going with my manual bisecting with the top level yt directories first to cut down on possibilities.

**Comment by chrishavlin** [comment]
Summary of this morning's manual bisecting:

```
pytest -v yt/data_objects yt/fields yt/frontends yt/geometry yt/sample_data yt/tests yt/units yt/utilities yt/visualization

	(full test suite) fails 


pytest -v yt/utilities yt/visualization

	passes

pytest --color=yes yt/sample_data yt/tests yt/units yt/utilities yt/visualization

	passes
	
pytest --color=yes yt/fields yt/frontends yt/geometry yt/sample_data yt/tests yt/units yt/utilities yt/visualization
	
	passes
	
pytest -v yt/data_objects yt/visualization

	fails!
```

**Comment by chrishavlin** [comment]
looking at what has changed recently in data_objects... my extrema-with-nans PR (#4762) **did** just get merged a few days ago. so latest push here reverts those changes just to check (that PR did pass tests prior to merging, but it also touches parallel objects...).

**Comment by chrishavlin** [comment]
unrelated to #4762 , reverted that. latest push further bisects test suite to just data_object/tests

**Comment by chrishavlin** [comment]
just moved a copy of the failing test to the `yt/data_objects/test` directory so i can try to use the detect-test-pollution plugin more easily (without running the entire test suite)

**Comment by chrishavlin** [comment]
oh interesting! detect-test-pollution returns:


```
discovering all tests...
-> discovered 174 tests!
ensuring test passes by itself...
-> OK!
ensuring test fails with test group...
-> OK!
running step 1:
- 173 tests remaining (about 8 steps)
running step 2:
- 86 tests remaining (about 7 steps)
running step 3:
- 43 tests remaining (about 6 steps)
running step 4:
- 22 tests remaining (about 5 steps)
running step 5:
- 11 tests remaining (about 4 steps)
running step 6:
- 6 tests remaining (about 3 steps)
running step 7:
- 3 tests remaining (about 2 steps)
running step 8:
- 2 tests remaining (about 1 steps)
double checking we found it...
-> the polluting test is: yt/data_objects/tests/test_firefly.py::test_firefly_JSON_string
```

not sure exactly how `detect-test-pollution` works, so this could be a false positive, but i'll look more closely...

**Comment by chrishavlin** [comment]
Summary:

```
        pytest --color=yes  \
        yt/visualization/volume_rendering/tests/test_scene.py \
        yt/data_objects/tests/test_firefly.py::test_firefly_JSON_string 
      
```
passes while
```
        pytest --color=yes  \
        yt/data_objects/tests/test_firefly.py::test_firefly_JSON_string \
        yt/visualization/volume_rendering/tests/test_scene.py
```
fails

**Comment by chrishavlin** [comment]
Still fails with just the creation of the firefly obj via `create_firefly_object`, so not write related (it wasn't writing anyway...)

**Comment by chrishavlin** [comment]
to reproduce locally, install Pyarrow!!!!!!!

**Comment by neutrinoceros** [comment]
I confirm I can reproduce locally with
```
pytest --color=yes  \
  yt/data_objects/tests/test_firefly.py::test_firefly_JSON_string \
  yt/visualization/volume_rendering/tests/test_scene.py
```
while the reverse order passes.
However it seems that `pyarrow` isn't needed to reproduce (unless it actually leaves traces after the first failure...) ? I need to check with a brand new env too.

**Comment by chrishavlin** [comment]
> However it seems that pyarrow isn't needed to reproduce

that would certainly make sense since I can't find any evidence of it being installed in the CI action. but something about installing it did let me reproduce the error locally... i'll try again in a fresh environment.

**Comment by neutrinoceros** [comment]
@chrishavlin https://github.com/yt-project/yt/issues/4785#issuecomment-1911680263 👀

**Comment by chrishavlin** [comment]
ya, I saw that -- but this branch is silencing those warnings already? In any case, I've now reproduced this locally without installing pyarrow. Just doing a minimal yt build and installing firefly

```
pip install -e . 
pip install firefly
```
then running the following script:

```python
from yt.testing import fake_random_ds
from yt import create_scene
from matplotlib.image import imread
from firefly import data_reader


def test_annotations(fname):
    ds = fake_random_ds(3)
    sc = create_scene(ds)
    sc.save_annotated(fname)
    image = imread(fname)
    print(f"image shape: {image.shape}")
    print(f"camera res: {sc.camera.resolution}")
    assert image.shape == sc.camera.resolution + (4,)


test_annotations("test_prefire.png") # runs as expected
reader = data_reader.Reader(
            datadir=None, clean_datadir=True
         )  
test_annotations("test_postfire.png") # fails at the assert statement
```
and checking the 2 saved files, the second one does indeed have different pixel dimensions.

**Comment by matthewturk** [comment]
I'll quick take a look but I'd bet it's related to modifying matplotlib parameters.

**Comment by matthewturk** [comment]
So I ran this, and got errors on the first test line.  I put this at the top:

```python
import matplotlib as mpl
mpl.rcParams.update(mpl.rcParamsDefault)
```

and the errors on the first line went away.  (My image size, like Chris's, was 614x614.)  I checked and the only thing different between my rcParams and the other is `figure.dpi` was set to 120, which would explain the exact difference.

So, why is it changing?  My guess or suspicion is that it's related to the current working directory, or some environment variable or something, but I'll check that in a moment.

**Comment by chrishavlin** [comment]
ya, this will show the change in dpi, which changes in the rcParams after the firefly import:

```python
import matplotlib as mpl
rcparams0 = mpl.rcParams.copy()
import firefly.data_reader
rcparams1 = mpl.rcParams.copy()

print(rcparams0['figure.dpi'])
print(rcparams1['figure.dpi'])
```
which prints 100 and 120

**Comment by chrishavlin** [comment]
OK, i think i have a fix on the yt side...

File: yt/visualization/volume_rendering/scene.py [hunk 1]
```diff
@@ -526,7 +526,9 @@ def _show_mpl(self, im, sigma_clip=None, dpi=100):
         from matplotlib.figure import Figure
 
         s = im.shape
-        self._render_figure = Figure(figsize=(s[1] / float(dpi), s[0] / float(dpi)))
+        self._render_figure = Figure(
+            figsize=(s[1] / float(dpi), s[0] / float(dpi)), dpi=dpi
```
Comment (chrishavlin) [inline yt/visualization/volume_rendering/scene.py:530]: note, this is the actual fix. the default figure dpi was changing due to the firefly import (a question for firefly folks?) and so this Figure needs to be instantiated with the `dpi` variable value

**Comment by chrishavlin** [comment]
assuming that this fix actually passes I'm going to close this PR and open a new one to avoid all the garbage commits here...

**Comment by chrishavlin** [comment]
Replaced by #4792 . Will close this one after I update the related issue summarizing everything.

**Comment by chrishavlin** [comment]
Ok, closing out now!

**Pull request closed without merging**